marshal.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// Marshal takes the input structure and creates an http request.
+//
+// See: Unmarshal for more details.
+//
+// For fields with a "path" item in the structural tag, the base uri must
+// contain a placeholder with its name.
+// Example:
+// For
+//    type Test struct {
+//	    username string `httprequest:"user,path"`
+//    }
+// ...the request url must contain a ":user" placeholder:
+//    http://localhost:8081/:user/files
+//
+// If a type does not implement the encoding.TextMarshaler fmt.Sprint will
+// be used to marshal its value.
+func Marshal(baseURL, method string, x interface{}) (*http.Request, error) {
+	xv := reflect.ValueOf(x)
+	pt, err := getRequestType(xv.Type())
+	if err != nil {
+		return nil, errgo.WithCausef(err, ErrBadUnmarshalType, "bad type %s", xv.Type())
+	}
+	req, err := http.NewRequest(method, baseURL, bytes.NewBuffer(nil))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	p := &Params{req, httprouter.Params{}}
+	if err := marshal(p, xv, pt); err != nil {
+		return nil, errgo.Mask(err, errgo.Is(ErrUnmarshal))
+	}
+	return p.Request, nil
+}
+
+// marshal is the internal version of Marshal.
+func marshal(p *Params, xv reflect.Value, pt *requestType) error {
+	if xv.Kind() == reflect.Ptr {
```

Comment:
This check is unnecessary - we know that it's a pointer because getRequestType checks that.